File: joyhousebot/api/routers/memory.py
"""Authenticated owner API for governed Memory updates."""

from __future__ import annotations

import asyncio
from typing import Literal

from fastapi import APIRouter, HTTPException, Query

from joyhousebot.api.dependencies import ContainerDep, ContextDep
from joyhousebot.api.schemas import ResolveMemoryCandidateRequest
from joyhousebot.application.presenters import record_dict

router = APIRouter(prefix="/memory", tags=["memory"])


@router.get("/documents")
async def list_memory_documents(
    context: ContextDep,
    container: ContainerDep,
    agent_id: str = Query(min_length=1, max_length=200),
    layer: Literal["profile", "long_term", "episodic", "agent", "all"] = "all",
    search: str | None = Query(default=None, max_length=200),
    limit: int = Query(default=200, ge=1, le=500),
):
    """List this user's durable Memory documents for one Agent."""
    items, summary = await asyncio.gather(
        asyncio.to_thread(
            container.store.list_memory_documents,
            user_id=context.user_id,
            agent_id=agent_id,
            layer=None if layer == "all" else layer,
            search=search,
            limit=limit,
        ),
        asyncio.to_thread(
            container.store.summarize_memory_documents,
            user_id=context.user_id,
            agent_id=agent_id,
        ),
    )
    return {"items": items, "summary": summary}


@router.get("/documents/{document_path:path}")
async def get_memory_document(
    document_path: str,
    context: ContextDep,
    container: ContainerDep,
    agent_id: str = Query(min_length=1, max_length=200),
    scope_key: str = Query(min_length=1, max_length=500),
):
    """Read one full document after matching user, Agent, scope and path."""
    item = await asyncio.to_thread(
        container.store.get_memory_document,
        user_id=context.user_id,
        agent_id=agent_id,
        scope_key=scope_key,
        document_path=document_path,
    )
    if item is None:
        raise HTTPException(status_code=404, detail="Memory document not found")
    return item


@router.get("/candidates")
async def list_memory_candidates(
    context: ContextDep,
    container: ContainerDep,
    agent_id: str | None = Query(default=None, min_length=1, max_length=200),
    status: Literal[
        "pending", "merged", "rejected", "expired", "conflicted", "all"
    ] = "pending",
    limit: int = Query(default=100, ge=1, le=500),
):
    rows = await container.memory_candidates.list(
        context,
        agent_id=agent_id,
        status=None if status == "all" else status,
        limit=limit,
    )
    return {"items": [record_dict(row) for row in rows]}


@router.post("/candidates/{candidate_id}/resolve")
async def resolve_memory_candidate(
    candidate_id: str,
    body: ResolveMemoryCandidateRequest,
    context: ContextDep,
    container: ContainerDep,
):
    candidate = await container.memory_candidates.resolve(
        context,
        candidate_id,
        resolution=body.resolution,
        note=body.note,
    )
    return record_dict(candidate)

File: apps/console/src/api/memory.ts
import { apiFetch } from './http'
import { getIdentityHeaders } from './identity'

export type MemoryLayer = 'profile' | 'long_term' | 'episodic' | 'agent'

export interface MemoryDocumentListItem {
  scope_key: string
  document_path: string
  layer: MemoryLayer
  version: number
  size_bytes: number
  preview: string
  created_at_ms: number
  updated_at_ms: number
}

export interface MemoryDocument extends Omit<MemoryDocumentListItem, 'preview'> {
  content: string
}

export interface MemoryDocumentSummary {
  total: number
  by_layer: Record<MemoryLayer, number>
}

export interface MemoryCandidate {
  candidate_id: string
  agent_id: string
  scope_key: string
  document_path: string
  layer: MemoryLayer
  operation: 'replace' | 'append'
  content: string
  source_run_id?: string | null
  source_kind: string
  fact_type: string
  confidence?: number | null
  data_classification: string
  status: 'pending' | 'merged' | 'rejected' | 'expired' | 'conflicted'
  created_at: string
  expires_at: string
  resolved_at?: string | null
}

async function memoryFetch<T>(path: string, init: RequestInit = {}): Promise<T> {
  const headers = new Headers(init.headers)
  for (const [key, value] of Object.entries(getIdentityHeaders())) headers.set(key, value)
  const response = await apiFetch(`/v1/memory${path}`, { ...init, headers })
  const payload = await response.json().catch(() => ({}))
  if (!response.ok) throw new Error(payload?.detail ?? payload?.error?.message ?? '记忆数据读取失败')
  return payload as T
}

export function getMemoryDocuments(
  agentId: string,
  filters: { layer?: MemoryLayer | 'all'; search?: string; limit?: number } = {},
) {
  const query = new URLSearchParams({ agent_id: agentId })
  if (filters.layer && filters.layer !== 'all') query.set('layer', filters.layer)
  if (filters.search) query.set('search', filters.search)
  query.set('limit', String(filters.limit ?? 200))
  return memoryFetch<{ items: MemoryDocumentListItem[]; summary: MemoryDocumentSummary }>(`/documents?${query}`)
}

export function getMemoryDocument(agentId: string, item: MemoryDocumentListItem) {
  const query = new URLSearchParams({ agent_id: agentId, scope_key: item.scope_key })
  const path = item.document_path.split('/').map(encodeURIComponent).join('/')
  return memoryFetch<MemoryDocument>(`/documents/${path}?${query}`)
}

export function getMemoryCandidates(
  agentId: string,
  status: MemoryCandidate['status'] | 'all' = 'all',
) {
  const query = new URLSearchParams({ agent_id: agentId, status, limit: '200' })
  return memoryFetch<{ items: MemoryCandidate[] }>(`/candidates?${query}`)
}

export function resolveMemoryCandidate(
  candidateId: string,
  resolution: 'accept' | 'reject',
  note?: string,
) {
  return memoryFetch<MemoryCandidate>(`/candidates/${encodeURIComponent(candidateId)}/resolve`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ resolution, note: note || null }),
  })
}
